lib/identity_cache/cache_fetcher.rb

```diff
@@ -24,27 +73,183 @@ def fetch_multi(keys, &block)
+            return data
+          else
+            raise LockWaitTimeout if lock_wait_limit <= 0
+            lock_wait_limit -= 1
+
+            # If fill failed in the other client, then it might be failing fast
+            # so avoid waiting the typical amount of time for a lock wait. The
+            # semian gem can be used to handle failing fast when the database is slow.
+            if lock.fill_failed?
+              return fetch_without_fill_lock(key) { yield }
+            end
+
+            # lock wait
+            sleep(fill_lock_duration)
+            # loop around to retry fetch_or_take_lock
+          end
+        when IdentityCache::DELETED # interrupted by cache invalidation
+          if lock && !using_fallback_key
+            # cache invalidated during lock wait, try fallback key
+            using_fallback_key = true
+            key = lock_fill_fallback_key(key, lock)
+            # Use lower TTL for fallback key lock since it won't be used for very long.
+            expires_in = fill_lock_duration * 2
+            # Memcached uses integer number of seconds for TTL so round up to avoid having
+            # the cache store round down with `to_i`.
+            expires_in = expires_in.ceil
+            # Memcached only uses part of the first second of the TTL, so add 1 to compensate.
+            expires_in += 1
+            next
+          else
+            # cache invalidation prevented lock from being taken
+            return yield
+          end
+        when nil # Errors talking to memcached
+          return yield
+        else # hit
+          return result
+        end
+      end
+      raise "unexpected number of loop iterations"
+    end
+
+    def cas_or_add(key, expires_in: nil)
+      yielded = false
+      swapped = @cache_backend.cas(key, expires_in: expires_in) do |value|
+        yielded = true
+        yield value
       end
       unless yielded
-        result = yield
-        add(key, result)
+        data = yield nil
+        swapped = add(key, data, expires_in: expires_in)
       end
-      result
+      swapped
     end
 
-    private
+    def fetch_or_take_lock(key, old_lock:, expires_in:)
+      new_lock = nil
+      swapped = cas_or_add(key, expires_in: expires_in) do |value|
+        if value.nil? || value == IdentityCache::DELETED
+          if old_lock # cache invalidated
+            return value
+          else
+            new_lock = FillLock.build(client_id: client_id, data_version: SecureRandom.uuid)
+          end
+        elsif FillLock.cache_value?(value)
+          if old_lock && old_lock.equals_cache_value?(value)
+            # preserve data version since there hasn't been any cache invalidations
+            new_lock = FillLock.build(client_id: client_id, data_version: old_lock.data_version)
+          else
+            return FillLock.from_cache(value)
+          end
+        else # hit
+          return value
+        end
+        new_lock.cache_value # take lock
```

Comment:
Currently the code convert the value to primitives so that they can be serialized and deserialized with msgpack.  However, an alternative would be to add a special value for `msgpack` to handle the hydration of these values instead of doing it explicitly in the code.  With `Marshal` as the serializer, it can handle this for us as long as we are careful about not changing the class name or instance variables names in an incompatible way.